dbms/src/Interpreters/RowRefs.h

```diff
@@ -32,14 +32,70 @@ struct RowRefList : RowRef
     RowRefList(const Block * block_, size_t row_num_) : RowRef(block_, row_num_) {}
 };
 
+/**
+ * This class is intended to push sortable data into.
+ * When looking up values the container ensures that it is sorted for log(N) lookup
+ * After calling any of the lookup methods, it is no longer allowed to insert more data as this would invalidate the
+ * references that can be returned by the lookup methods
+ */
+
+template <typename T>
+class SortedLookupVector
+{
+public:
+    using Base = std::vector<T>;
+
+    // First stage, insertions into the vector
+    template <typename U, typename ... TAllocatorParams>
+    void insert(U && x, TAllocatorParams &&... allocator_params)
+    {
+        assert(!sorted.load(std::memory_order_acquire));
+        array.push_back(std::forward<U>(x), std::forward<TAllocatorParams>(allocator_params)...);
+    }
+
+    // Transition into second stage, ensures that the vector is sorted
+    typename Base::const_iterator upper_bound(const T & k)
+    {
+        sort();
+        return std::upper_bound(array.cbegin(), array.cend(), k);
+    }
+
+    // After ensuring that the vector is sorted by calling a lookup these are safe to call
+    typename Base::const_iterator cbegin() const { return array.cbegin(); }
+    typename Base::const_iterator cend() const { return array.cend(); }
+
+private:
+    Base array;
+    std::atomic<bool> sorted = false;
```

Comment:
Probably meaningless but won't we place atomic the first element in the class to make sure it's well-aligned and fetched regardless sizeof(Base)?

Reply:
I don't think it's going to make any difference (as it isn't contending with any writes nearby after the insertion), but I'll happily move it